assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js

```diff
@@ -0,0 +1,395 @@
+			.select( MODULES_ANALYTICS_4 )
+			.selectCustomDimensionDataAvailability( customDimension );
+
+		yield baseActions.receiveIsCustomDimensionGatheringData(
+			customDimension,
+			! dataAvailability
+		);
+
+		if ( dataAvailability ) {
+			yield fetchSaveCustomDimensionDataAvailableStateStore.actions.fetchSaveCustomDimensionDataAvailableState(
+				customDimension
+			);
+		}
+	},
+};
+
+const baseSelectors = {
+	/**
+	 * Determines whether data is available for the custom dimension.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state           Data store's state.
+	 * @param {string} customDimension Custom dimension slug.
+	 * @return {boolean|undefined|null} Returns TRUE if data is available, otherwise FALSE.
+	 *                                  If the request is still being resolved, returns undefined.
+	 *                                  If the data availability can not be determined, returns null.
+	 */
+	selectCustomDimensionDataAvailability: createRegistrySelector(
+		( select ) => ( state, customDimension ) => {
+			if (
+				! select( MODULES_ANALYTICS_4 ).hasCustomDimensions(
+					customDimension
+				)
+			) {
+				return false;
+			}
+
+			const isAuthenticated = select( CORE_USER ).isAuthenticated();
+
+			if ( isAuthenticated === undefined ) {
+				return undefined;
+			}
+			if ( ! isAuthenticated ) {
+				return false;
+			}
+
+			const propertyID = select( MODULES_ANALYTICS_4 ).getPropertyID();
+
+			if ( propertyID === undefined ) {
+				return undefined;
+			}
+
+			const property =
+				select( MODULES_ANALYTICS_4 ).getProperty( propertyID );
+
+			if ( property === undefined ) {
+				return undefined;
+			}
+
+			const startDate = getDateString( new Date( property.createTime ) );
+
+			const endDate = select( CORE_USER ).getReferenceDate();
+
+			const reportArgs = {
+				startDate,
+				endDate,
+				dimensions: [ `customEvent:${ customDimension }` ],
+			};
+
+			// We may legitimately want to return early before using `report` if the report is not resolved or results in an error.
+			// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+			const report =
+				select( MODULES_ANALYTICS_4 ).getReport( reportArgs );
+
+			const hasResolvedReport = select(
+				MODULES_ANALYTICS_4
+			).hasFinishedResolution( 'getReport', [ reportArgs ] );
+
+			if ( ! hasResolvedReport ) {
+				return undefined;
+			}
+
+			const hasReportError = select(
+				MODULES_ANALYTICS_4
+			).getErrorForSelector( 'getReport', [ reportArgs ] );
+
+			// If there is an error, return `null` since we don't know if there is data or not.
+			if ( hasReportError ) {
+				return null;
+			}
+
+			if ( ! report?.rows?.length ) {
+				return false;
+			}
+
+			// If the only dimension value is '(not set)', then there is no data. See https://support.google.com/analytics/answer/13504892.
+			const isZeroReport =
+				report.rowCount === 1 &&
+				report.rows[ 0 ].dimensionValues?.[ 0 ]?.value === '(not set)';
```

Comment:
I think we should also consider report results that have no rows as zero report as well, right? In my testing I saw reports with zero rows for custom dimensions.

Reply:
I've already got an explicit check for the "no rows" case above this, as I didn't want to overcomplicate the condition here. I'd be happy to merge it with this condition though, if you like?

Reply:
Ah, yes, you are right. Missed that condition. Looks good to me then.